sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/generator/AbstractQueryGenerator.java

```diff
@@ -268,9 +268,26 @@ protected SqlQuerySpec generateCosmosQuery(@NonNull CosmosQuery query,
                                                          toCosmosDbValue(p.getSecond())))
                                                      .collect(Collectors.toList());
 
-        if (query.getLimit() > 0) {
+
+        long offset = 0;
+        long pageSize = 0;
+        try {
+                offset = query.getPageable().getOffset();
+                pageSize = query.getPageable().getPageSize();
+        } catch (UnsupportedOperationException ex) {
+            // Ignore
+        }
+
+        if (offset > 0 && pageSize > 0 && !queryString.contains("COUNT")) {
+            queryString = new StringBuilder(queryString)
+                .append(" OFFSET ")
+                .append(offset)
+                .append(" LIMIT ")
+                .append(pageSize).toString();
```

Comment:
since both offset and limit count are queried in the offset limit cause, so I kind feel we should just use whatever the limit as the limit, but I am not sure in Spring world, whether it makes sense @mbhaskar @kushagraThapar

Reply:
When you calculate the offset, it needs to actually be the page size * page requested to generate the correct offset. For example, if someone is asking for 50 items of the first page, you would have a page=0, size=50 so offset=50*0=0 and limit=50. Second page is page=1 so offset=50*1=50 and limit=50. 

We achieved this using a helper method before building the query:
```
fun toPageRequestWithOffset(defaultSort: Sort) : PageRequest {
        return CosmosPageRequest.of((size * page).toLong() ,page, size, token, calculateSort(defaultSort))
    }
```
This built a CosmosPageRequest with the math already handled so that it would be passed cleanly into the query generator call. 

Not having this baked into the Azure Cosmos code and forcing the consumers to do that before passing it in would be less than ideal IMO.